lib/Sema/CodeSynthesis.cpp

```diff
@@ -2192,28 +2193,65 @@ ConstructorDecl *swift::createImplicitConstructor(TypeChecker &tc,
       // If var is a lazy property, its value is provided for the underlying
       // storage.  We thus take an optional of the properties type.  We only
       // need to do this because the implicit constructor is added before all
-      // the properties are type checked.  Perhaps init() synth should be moved
+      // the properties are type checked. Perhaps init() synth should be moved
       // later.
       if (var->getAttrs().hasAttribute<LazyAttr>())
         varInterfaceType = OptionalType::get(varInterfaceType);
 
       // Create the parameter.
-      auto *arg = new (context)
+      auto *arg = new (C)
           ParamDecl(VarDecl::Specifier::Default, SourceLoc(), Loc,
                     var->getName(), Loc, var->getName(), decl);
       arg->setInterfaceType(varInterfaceType);
       arg->setImplicit();
       
+      // If this is a var that has a default value, lets assign a default value
+      // to the parameter with the same expression.
+      if (!var->isLet()) {
+        if (auto init = var->getParentInitializer()) {
+          // Give this some bogus context right now, we'll fix it after making
+          // the constructor.
+          auto *initDC = new (C) DefaultArgumentInitializer(
+            arg->getDeclContext(), params.size());
+            
+          defaultInits.push_back(initDC);
+          
+          // Set the default value to the variable. When we emit this in silgen
+          // we're going to call the variable's initializer expression.
+          arg->setStoredProperty(var);
+          arg->setDefaultArgumentKind(DefaultArgumentKind::StoredProperty);
+        }
+      }
+      
+      // Now that we have default values for this synthesized constructor,
+      // if the property is an optional and does not have an initializer,
+      // and is not a let property because it can never be reassigned,
+      // assign nil as the default value.
+      if (var->getType()->getOptionalObjectType() &&
+          var->getParentPatternBinding()->isDefaultInitializable() &&
+          !var->isLet() &&
+          !var->getParentInitializer()) {
+        auto *initDC = new (C) DefaultArgumentInitializer(
+          arg->getDeclContext(), params.size());
+
+        defaultInits.push_back(initDC);
+        
+        auto nil = new (C) NilLiteralExpr(SourceLoc(), /*implicit*/ true);
```

Comment:
I don't think we can because given:
```swift
struct X {
  var y: Int?
  var z = 0
}

let x = X(z: 16)
```
Type checker catches this before this is handed off to SILGen because memberwise initializer produced is:
```swift
init(y: Int?, z: Int = z)
```
Which produces the error: `cannot invoke initializer for type 'X' with an argument list of type '(z: Bool)'`.

Reply:
I'm still not sure you need to actually build the NilLiteralExpr here and set the default argument kind to NilLiteral. Since the argument is emitted in SILGen as a call to the property initializer, you should still be able to use DefaultArgumentKind::StoredProperty here and avoid building the NilLiteralExpr.